### PR TITLE
Add software blinking to GPIO LEDs

### PR DESCRIPTION
**Pull request by blemouzy**
- Add software blinking to GPIO LEDs
- Add missing `blink` command to led shell

### CODE REVIEW

File: drivers/led/Kconfig.gpio [hunk 1]
```diff
@@ -7,3 +7,23 @@ config LED_GPIO
...
+	help
+	  Enable software LED blinking.
+
+if LED_GPIO_BLINK
+
+config LED_GPIO_BLINK_NUM
+	int "Number of LEDs to manage"
+	default 4
+	help
+	  Number of LEDs to manage.
```
Comment (gmarull) [inline drivers/led/Kconfig.gpio:25]: you now the number of children at compile time, so why not use that information?

**Review by gmarull — CHANGES_REQUESTED** [review]

File: drivers/led/Kconfig.gpio (hunk 1 above)
Comment (blemouzy) [reply drivers/led/Kconfig.gpio:25]: How can I retrieve it?
I can't use `ARRAY_SIZE(gpio_dt_spec_##i)` in `struct led_gpio_data`.

Comment (blemouzy) [reply drivers/led/Kconfig.gpio:25]: I found an example in `drivers/led/lp503x.c` to do that.
Thanks for the comment.

Comment (blemouzy) [reply drivers/led/Kconfig.gpio]: Fixed

File: drivers/led/led_gpio.c [hunk 2]
```diff
@@ -49,16 +63,74 @@ static int led_gpio_off(const struct device *dev, uint32_t led)
...
+		CONTAINER_OF(timer, struct led_gpio_blink_data, timer);
+	uint32_t led = ARRAY_INDEX(data->blink, blink_data);
+	const struct gpio_dt_spec *led_gpio;
+
+	if (led >= config->num_leds) {
+		return;
+	}
+
+	led_gpio = &config->led[led];
+	if (gpio_pin_get_dt(led_gpio)) {
```
[the remark this answers is not in the document]
Comment (blemouzy) [reply drivers/led/led_gpio.c]: Fixed
The current status of LEDs are now stored in driver data.

**Review by simonguinot — CHANGES_REQUESTED** [review]
Hi @blemouzy,

Thanks for this PR but I can't agree with this.

The plan for software LED blinking is to implement a fallback at the LED core level. This way it will be available for ALL the LED drivers and not only led_gpio.

If a driver implements the blink API function (because the feature is supported by hardware) then it can be used directly else we can fall back on a software blinking mechanism (based on timer and/or workqueue).

For reference see the following discussions:

- https://github.com/zephyrproject-rtos/zephyr/pull/54594#discussion_r1099954315
- https://github.com/zephyrproject-rtos/zephyr/pull/50511 (for color correction but the idea is the same)

**Pull request closed without merging**